src/librustc_typeck/check/mod.rs

```diff
@@ -819,14 +819,10 @@ pub fn check_item_body<'a,'tcx>(ccx: &CrateCtxt<'a,'tcx>, it: &'tcx hir::Item) {
                     check_const(ccx, trait_item.span, &expr, trait_item.id)
                 }
                 hir::MethodTraitItem(ref sig, Some(ref body)) => {
```

Comment:
can you add an assertion here that the fn is not const? (and maybe a comment pointing out that this is checked in the AST validator?)